doc/PROPOSAL.md

```diff
@@ -0,0 +1,163 @@
+so that such events are stored across multiple database key/value pairs. To
+achieve event fragmentation, key/value pairs will be formatted like so:
+
+#### Keys/Values
+
+|           | encoding                                                                           | format                         |
+|-----------|------------------------------------------------------------------------------------|--------------------------------|
+| **key**   | `0x00` &#x7c; 64-bit event id (big-endian) &#x7c; 32-bit fragment id (big-endian)  | `0x00112233445566778899AABBCC` |
+| **value** | raw bytes starting at index `(i * <optimal fragment size>)` for fragment w/ id `i` | `0xdeadbeef...`                |
+
+Notes:
+- `|` = concatenation
+- The event ID and fragment ID are both contiguous, non-negative integers
+- The optimal fragment size, as documented above, will almost certainly be
+  `10 KB` = `10,000 bytes`
+- The first fragment (fragment `0`) will have a slightly different encoding for
+  the value, as documented below
+
+#### Fragment 0
+
+The value of the first fragment of an event will be split into two parts that
+are concatenated: the "header" and the "payload".
+
+#### Fragment 0 Header
+
+The fragment header is used to record the total number of fragments for the
+event. This serves two functions: it allows the read algorithm to pre-allocate
+an appropriate amount of memory on the heap for an event after reading the
+first batch out of FDB, and `memcpy` data from FDB as it arrives in batches; it
+allows the read algorithm to verify that the number of fragments returned by FDB
+is the number of fragments expected.
+
+The fragment header is a raw byte array with the following format:
+
+**128 fragments or less**
+
+| byte | meaning                | format          | example     |
+|------|------------------------|-----------------|-------------|
+| 0    | *additional* fragments | `<# fragments>` | `0111 1111` |
+
+**129 fragments or less**
+
+| byte | meaning                      | format                                             | example      |
+|------|------------------------------|----------------------------------------------------|--------------|
+| 0    | additional bytes             | `0x80` &#x7c;&#x7c; `<bytes to store # fragments>` | `1000 0010 ` |
+| 1..N | bytes encoding `# fragments` | `<# fragments>`                                    | `0001 0110 ` |
+
+Note: `||` = bitwise OR
+
+Let's see how the 257th fragment of an example event which requires 257
+fragments to be stored would look:
+
+| byte | bucket        | format                                 | example     |
+|------|---------------|----------------------------------------|-------------|
+| 0    | byte-width    | overflow toggle bit, bucket byte-width | `1000 0010` |
+| 1    | `# fragments` | total fragments                        | `0000 0000` |
+| 2    | `# fragments` | total fragments                        | `0000 0001` |
```

Comment:
If this is supposed to encode the *additional* number of fragments left in the event, and this is an example of the last fragment in the sequence, shouldn't the example data read `0000 0000`?

Also perhaps we should replace "total fragments" with "remaining fragments". I actually like "remaining" better than "additional" because it more strongly implies a dynamic counter. Thoughts?

Reply:
I think you're misunderstanding. Only the first fragment, i.e. fragment 0, has a header. This header includes the number of additional/remaining fragments. Every other fragment has no header at all.

Re: additional vs. remaining - whichever you prefer

Reply:
Ah, got it. Ok it would be great to make that more apparent, with a sentence saying exactly what you said: "Only fragment 0 has a header, the rest have none at all."